main/modes/quickSettings/menuQuickSettingsRenderer.c

```diff
@@ -0,0 +1,166 @@
+#include "menuQuickSettingsRenderer.h"
+#include "menu_utils.h"
+
+#include "hdw-tft.h"
+#include "shapes.h"
+#include "fill.h"
+#include "wsg.h"
+#include "font.h"
+#include "menu.h"
+#include "palette.h"
+#include "esp_log.h"
+
+#include <malloc.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+//==============================================================================
+// Defines
+//==============================================================================
+
+/// The width of the panel, centered in the screen
+#define PANEL_W (TFT_WIDTH - 60)
+
+/// The height of the panel, from the top of the screen
+#define PANEL_H 60
+
+/// The panel corner-radius
+#define PANEL_R 15
+
+#define TEXT_MARGIN 10
+
+#define PANEL_BG_COLOR   c333
+#define PANEL_BOX_COLOR  c000
+#define PANEL_TEXT_COLOR c000
+
+#define ICON_W      16
+#define ICON_H      16
+#define ICON_BOX_PX 2
+
+//==============================================================================
+// Structs
+//==============================================================================
+
+typedef struct
+{
+    /// @brief The menu item label these
+    const char* label;
+    /// @brief The icon to draw when the setting is on
+    const wsg_t* onWsg;
+    /// @brief The icon to draw when the setting is off
+    const wsg_t* offWsg;
+} quickSettingsItemInfo_t;
+
+//==============================================================================
+// Static Function Prototypes
+//==============================================================================
+
+static const quickSettingsItemInfo_t* getInfoForLabel(menuQuickSettingsRenderer_t* renderer, const char* label);
+
+//==============================================================================
+// Static Functions
+//==============================================================================
+
+static const quickSettingsItemInfo_t* getInfoForLabel(menuQuickSettingsRenderer_t* renderer, const char* label)
+{
+    node_t* node = renderer->iconMap.first;
+
+    while (node != NULL)
+    {
+        quickSettingsItemInfo_t* info = (quickSettingsItemInfo_t*)(node->val);
+
+        // The labels should be the exact same strings, so it's fine to compare like this
+        if (info->label == label)
+        {
+            return info;
+        }
+
+        node = node->next;
+    }
+
+    return NULL;
+}
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+menuQuickSettingsRenderer_t* initMenuQuickSettingsRenderer(font_t* font)
+{
+    menuQuickSettingsRenderer_t* renderer = calloc(1, sizeof(menuQuickSettingsRenderer_t));
+
+    renderer->font = font;
+    return renderer;
+}
+
+void deinitMenuQuickSettingsRenderer(menuQuickSettingsRenderer_t* renderer)
+{
+    clear(&renderer->iconMap);
```

Comment:
I think this is a leak. `clear()` doesn't `free()` values in the linked list, it only frees the nodes. There's no guarantee the pointers `push()`'d to the list were dynamically allocated.

see the `callco()`:
```c
void quickSettingsRendererAddIcon(menuQuickSettingsRenderer_t* renderer, const char* label, const wsg_t* onWsg,
                                  const wsg_t* offWsg)
{
    quickSettingsItemInfo_t* info = calloc(1, sizeof(quickSettingsItemInfo_t));

    info->label  = label;
    info->onWsg  = onWsg;
    info->offWsg = offWsg;

    push(&renderer->iconMap, info);
}
```

Reply:
Yeah, confirmed by `drmemory.exe`. I wonder why `libasan` didn't whine about it?
```
Error #155: LEAK 24 direct bytes 0x000001f7907da7b0-0x000001f7907da7c8 + 0 indirect bytes
# 0 replace_calloc                             [d:\a\drmemory\drmemory\common\alloc_replace.c:2620]
# 1 quickSettingsRendererAddIcon               [main/modes/quickSettings/menuQuickSettingsRenderer.c:159]
# 2 quickSettingsEnterMode                     [main/modes/quickSettings/quickSettings.c:187]
# 3 app_main                                   [main/swadge2024.c:423]
# 4 main                                       [emulator/src/emu_main.c:104]

Error #156: LEAK 24 direct bytes 0x000001f7907da830-0x000001f7907da848 + 0 indirect bytes
# 0 replace_calloc                             [d:\a\drmemory\drmemory\common\alloc_replace.c:2620]
# 1 quickSettingsRendererAddIcon               [main/modes/quickSettings/menuQuickSettingsRenderer.c:159]
# 2 quickSettingsEnterMode                     [main/modes/quickSettings/quickSettings.c:189]
# 3 app_main                                   [main/swadge2024.c:423]
# 4 main                                       [emulator/src/emu_main.c:104]

Error #157: LEAK 24 direct bytes 0x000001f7907da8b0-0x000001f7907da8c8 + 0 indirect bytes
# 0 replace_calloc                             [d:\a\drmemory\drmemory\common\alloc_replace.c:2620]
# 1 quickSettingsRendererAddIcon               [main/modes/quickSettings/menuQuickSettingsRenderer.c:159]
# 2 app_main                                   [main/swadge2024.c:423]
# 3 main                                       [emulator/src/emu_main.c:104]
```

Reply:
Should be fixed now! Good catch